### ISSUE TITLE
@action throws invariant failed at runtime

### ISSUE THREAD

**Issue by otbe** [mobxjs/mobx#269]
Hi,

```@action``` does not work for me in my current project. It throws ```action can only be invoked on functions``` at runtime.

![bildschirmfoto 2016-05-27 um 13 33 12](https://cloud.githubusercontent.com/assets/3391052/15609931/05c938a6-2423-11e6-9e2a-3436d6b593c6.png)

Usage:
```javascript
class Test {
   @action
   private func = () => {}
}
```

mobx@2.2.1 and ts@1.8.10
TSC target is es6 with ```relflect-metadata``` package.

Project is available at github: https://github.com/otbe/dwatch/tree/dep-updates (https://github.com/otbe/dwatch/blob/dep-updates/app/src/components/images/Images.tsx#L179 with ```@action```)

I run a debug session and this are my results:
It seems that the arguments for https://github.com/mobxjs/mobx/blob/master/src/core/action.ts#L24 are swapped in my case.

<img width="794" alt="bildschirmfoto 2016-05-27 um 15 40 51" src="https://cloud.githubusercontent.com/assets/3391052/15609628/6eb6ce34-2421-11e6-92bf-e1f668c7dbc9.png">

<img width="798" alt="bildschirmfoto 2016-05-27 um 15 22 06" src="https://cloud.githubusercontent.com/assets/3391052/15609580/21e247aa-2421-11e6-8bc4-5d1c5ce8e053.png">

I think the order how the arguments are passed is fine, because [this](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Decorators.md#method-decorators) and [this](https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Decorators.md#accessor-decorators). But why mobx assumes something different?

Another problem: If the problem above is fixed it wont worker, because ```fn``` is not a function. Its the target (prototype) itself. This seems to be fine because the typescript implementation of ```decorate``` is (target es6):
```javascript
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
/*....*/
__decorate([_mobx.action, __metadata('design:type', Object)], Images.prototype, "changeFilter", void 0);
```
Where ```Images.prototype.changeFilter``` would be the function to decorate.

Grateful for advices. Thanks! :)

P.S.: I think its unrelated to ```reflect-metadata``` because the method signature of [```Reflect.decorate```](https://github.com/rbuckton/ReflectDecorators/blob/master/Reflect.ts#L174) matches the one expected by TS in:
```javascript
var __decorate = undefined && undefined.__decorate || function (decorators, target, key, desc) {
	    var c = arguments.length,
	        r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc,
	        d;
	    if ((typeof Reflect === "undefined" ? "undefined" : _typeof(Reflect)) === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);else for (var i = decorators.length - 1; i >= 0; i--) {
	        if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
	    }return c > 3 && r && Object.defineProperty(target, key, r), r;
	};
```

**Comment by mweststrate** [comment]
Can you verify this issue is fixed in 2.2.2?

**Comment by otbe** [comment]
Works now. Thanks!
